File: 0386-lexicographical-numbers/0386-lexicographical-numbers.java
class Solution {
    private List<Integer> ans = new LinkedList<>();
    private void genNums(int curr, int lim) {
        if (curr > lim) return;

        ans.add(curr);

        for (int next = 0; next <= 9; ++next) {
            int nextNum = curr * 10 + next;
            if (nextNum <= lim) genNums(nextNum, lim);
            else break;            
        }
    }
    public List<Integer> lexicalOrder(int n) {
        for (int st = 1; st <= 9; ++st) genNums(st, n);
        return ans;
    }
}

File: 0386-lexicographical-numbers/0386-lexicographical-numbers.kt
class Solution {
    fun lexicalOrder(n: Int): List<Int> {
        val ans: MutableList<Int> = mutableListOf();
        var curr: Int = 1;

        repeat(n) {
            ans.add(curr);

            if (curr * 10 <= n) curr *= 10;
            else {
                while (curr % 10 == 9 || curr >= n) curr /= 10;
                curr++;
            }
        }

        return ans;
    }
}